zkevm-circuits/src/evm_circuit/util/memory_gadget.rs

```diff
@@ -1,57 +1,166 @@
-use crate::evm_circuit::{
-    param::MAX_MEMORY_SIZE_IN_BYTES,
-    util::{
-        constraint_builder::ConstraintBuilder,
-        math_gadget::{ConstantDivisionGadget, MaxGadget},
-        Address, MemorySize,
+use std::convert::TryInto;
+
+use crate::{
+    evm_circuit::{
+        param::{N_BYTES_GAS, N_BYTES_MEMORY_ADDRESS, N_BYTES_MEMORY_SIZE},
+        util::{
+            constraint_builder::ConstraintBuilder,
+            from_bytes,
+            math_gadget::IsZeroGadget,
+            math_gadget::{ConstantDivisionGadget, MinMaxGadget},
+            sum, Cell, MemoryAddress, Word,
+        },
     },
+    util::Expr,
+};
+use array_init::array_init;
+use bus_mapping::{
+    eth_types::{ToLittleEndian, U256},
+    evm::GasCost,
 };
-use crate::util::Expr;
-use bus_mapping::evm::GasCost;
 use halo2::plonk::Error;
 use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Expression};
 
 /// Decodes the usable part of an address stored in a Word
 pub(crate) mod address_low {
     use crate::evm_circuit::{
-        param::NUM_ADDRESS_BYTES_USED,
-        util::{from_bytes, Address, Word},
+        param::N_BYTES_MEMORY_ADDRESS,
+        util::{from_bytes, Word},
     };
     use halo2::{arithmetic::FieldExt, plonk::Expression};
 
     pub(crate) fn expr<F: FieldExt>(address: &Word<F>) -> Expression<F> {
-        from_bytes::expr(&address.cells[0..NUM_ADDRESS_BYTES_USED])
+        from_bytes::expr(&address.cells[..N_BYTES_MEMORY_ADDRESS])
     }
 
-    pub(crate) fn value<F: FieldExt>(address: [u8; 32]) -> Address {
-        from_bytes::value::<F>(&address[0..NUM_ADDRESS_BYTES_USED])
-            .get_lower_128() as Address
+    pub(crate) fn value<F: FieldExt>(address: [u8; 32]) -> u64 {
+        from_bytes::value::<F>(&address[..N_BYTES_MEMORY_ADDRESS])
+            .get_lower_128() as u64
```

Comment:
How do you know that this will not fail? Since `get_lower_128()` returns the lower 128 bits. 
See: https://github.com/appliedzkp/pairing/blob/3f668ba2eab444ad9f4c6396db8292fada6feb42/src/arithmetic/fields.rs#L117

Therefore this should be able to fail.

Reply:
I was assuming the witness is already valid, to escape the `Option` or `Result` for such small function. There are many other places I use such methods which don't panic.

Reply:
Just realized the bytes are already capped to 5 bytes, so it should not fail. But I fixed it by using `u64::from`, to avoid such confusing conversion.